src/main/java/dev/morling/onebrc/CalculateAverage_thomaswue.java

```diff
@@ -114,44 +113,44 @@ private static Unsafe initUnsafe() {
         }
     }
 
-    private static void parseLoop(long chunkStart, long chunkEnd, Result[] results, HashMap<String, Result> cities) {
+    private static void parseLoop(long chunkStart, long chunkEnd, HashMap<String, Result> cities) {
+        Result[] results = new Result[1 << 18];
         long scanPtr = chunkStart;
         byte b;
         while (scanPtr < chunkEnd) {
             long nameAddress = scanPtr;
-            int hash = 0;
-
-            // Skip first letter.
-            scanPtr++;
-
-            // Scan for ';' delimiter, always 4 bytes at a time.
-            while (true) {
-                int nextVal = UNSAFE.getInt(scanPtr);
-                if ((nextVal & 0x3B) == 0x3B) {
-                    scanPtr++;
-                    break;
-                }
-                else if ((nextVal & 0x3B00) == 0x3B00) {
-                    scanPtr += 2;
-                    hash = hash ^ (nextVal & 0xFF);
-                    break;
-                }
-                else if ((nextVal & 0x3B0000) == 0x3B0000) {
-                    scanPtr += 3;
-                    hash = hash ^ (nextVal & 0xFFFF);
-                    break;
-                }
-                else if (((nextVal & 0x3B000000) == 0x3B000000)) {
-                    scanPtr += 4;
-                    hash = hash ^ (nextVal & 0xFFFFFF);
-                    break;
+            long hash = 0;
+
+            // Search for ';', one long at a time.
+            long word = UNSAFE.getLong(scanPtr);
```

Comment:
This read can go over the end of the file. For example, a file contains the last entry "X;0.0\n". The read goes over 2 bytes at the end of the file. It will touch the next page, if the file has size equal to page size. If the next page is protected, that is seg fault. But I can't reproduce it locally. I always have ~136K memory after the end of the mapped region which I can access witout getting a seg fault. Maybe you have an explaination? 

A workaround is to simply copy the tail with padding (equal has the same flaw and we want it read words).

Reply:
Yes, didn't think about that. It is very unlikely to exactly have a page boundary there, but will have to add the workaround for that last row.